libs/pubsub/query/query.go

```diff
@@ -169,86 +191,114 @@ func (q *Query) Matches(events map[string][]string) bool {
 		return false
 	}
 
-	buffer, begin, end := q.parser.Buffer, 0, 0
+	var (
+		eventAttr string
+		op        Operator
+	)
 
-	var tag string
-	var op Operator
+	buffer, begin, end := q.parser.Buffer, 0, 0
 
 	// tokens must be in the following order:
-	// tag ("tx.gas") -> operator ("=") -> operand ("7")
+	// event attribute ("tx.gas") -> operator ("=") -> operand ("7")
 	for _, token := range q.parser.Tokens() {
 		switch token.pegRule {
-
 		case rulePegText:
 			begin, end = int(token.begin), int(token.end)
+
 		case ruletag:
-			tag = buffer[begin:end]
+			eventAttr = buffer[begin:end]
+
 		case rulele:
 			op = OpLessEqual
+
 		case rulege:
 			op = OpGreaterEqual
+
 		case rulel:
 			op = OpLess
+
 		case ruleg:
 			op = OpGreater
+
 		case ruleequal:
 			op = OpEqual
+
 		case rulecontains:
 			op = OpContains
+
 		case rulevalue:
 			// strip single quotes from value (i.e. "'NewBlock'" -> "NewBlock")
 			valueWithoutSingleQuotes := buffer[begin+1 : end-1]
 
-			// see if the triplet (tag, operator, operand) matches any tag
+			// see if the triplet (event attribute, operator, operand) matches any event
 			// "tx.gas", "=", "7", { "tx.gas": 7, "tx.ID": "4AE393495334" }
-			if !match(tag, op, reflect.ValueOf(valueWithoutSingleQuotes), events) {
+			if !match(eventAttr, op, reflect.ValueOf(valueWithoutSingleQuotes), events) {
 				return false
 			}
+
 		case rulenumber:
 			number := buffer[begin:end]
 			if strings.ContainsAny(number, ".") { // if it looks like a floating-point number
 				value, err := strconv.ParseFloat(number, 64)
 				if err != nil {
-					panic(fmt.Sprintf(
-						"got %v while trying to parse %s as float64 (should never happen if the grammar is correct)",
+					fmt.Printf(
```

Comment:
Why panic here and not in `matchValue`? Why panic at all in query-handling logic?

Reply:
I am agreeing with bez on this one, I can think of a reason to panic here, the incorrect grammar will be noticed by the developer and fixed.

Reply:
error here indicates a fundamental error with the grammar, which I'd rather be aware immediately. I.e. it's a corner case, not a normal situation.

Reply:
maybe we should not panic, but return an error...

Reply:
note: it will only panic if Query.New("foo.bar > 50") returns OK, but Conditions suddenly discovers that "foo.bar > 50" is invalid, which should not happen NORMALLY if parser (Query.New step) works correctly